### PR TITLE
Add stall detection to chaos test.

### PR DESCRIPTION
**Pull request by BramGruneir**
[Review on Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/3300)

### CODE REVIEW

File: acceptance/chaos_test.go [hunk 1]
```diff
@@ -145,13 +159,33 @@ func TestChaos(t *testing.T) {
...
-			// running.
-			cur := make([]string, num)
-			for i := range cur {
-				cur[i] = fmt.Sprintf("%d", atomic.LoadInt64(&counts[i]))
+			var newOutput string
+			if time.Now().Before(deadline) {
+				curRound := atomic.LoadInt64(&roundCopy)
+				if curRound == oldRound {
+					if time.Now().After(stallTime) {
+						close(stopper)
```
Comment (tschottdorf) [inline acceptance/chaos_test.go]: you're technically not supposed to close `stopper` (it's racy, though a very unlikely race).

File: acceptance/chaos_test.go [hunk 2]
```diff
@@ -113,27 +115,39 @@ func TestChaos(t *testing.T) {
 				return
 			default:
 			}
+			atomic.StoreInt64(&roundCopy, int64(round))
```
Comment (tschottdorf) [inline acceptance/chaos_test.go]: just use atomics to update `round` directly (which can be a test-global variable).

File: acceptance/chaos_test.go [hunk 3]
```diff
@@ -113,27 +115,39 @@ func TestChaos(t *testing.T) {
...
 				c.Kill(i)
 				c.Restart(i)
 				initClient(i)
+			}
+			for i := 0; i < num; i++ {
 				clients[i].Unlock()
 			}
-			for cur := atomic.LoadInt64(&count); time.Now().Before(deadline) &&
-				atomic.LoadInt64(&count) == cur; time.Sleep(time.Second) {
+			for cur := atomic.LoadInt64(&count); time.Now().Before(deadline) && atomic.LoadInt64(&count) == cur; time.Sleep(time.Second) {
```
Comment (tamird) [inline acceptance/chaos_test.go]: undo this change

File: acceptance/chaos_test.go [hunk 4]
```diff
@@ -113,27 +115,39 @@ func TestChaos(t *testing.T) {
...
+				}
+				if time.Now().After(deadline) {
+					break
+				}
+				log.Infof("restarting %d", i)
 				c.Kill(i)
 				c.Restart(i)
 				initClient(i)
+			}
+			for i := 0; i < num; i++ {
```
Comment (tamird) [inline acceptance/chaos_test.go]: why do you need this separate iteration?

File: acceptance/chaos_test.go (hunk 2 above)
Comment (tamird) [inline acceptance/chaos_test.go]: pretty sure you don't need the `int64()` here since `round` is initialized to an untyped constant

File: acceptance/chaos_test.go (hunk 1 above)
Comment (BramGruneir) [inline acceptance/chaos_test.go]: After discussing this offline, I should be able to remove the close.

File: acceptance/chaos_test.go (hunk 4 above)
Comment (BramGruneir) [inline acceptance/chaos_test.go]: This was to ensure that if we exit the the previous loop, that we will still release all locks.

File: acceptance/chaos_test.go (hunk 2 above)
Comment (BramGruneir) [inline acceptance/chaos_test.go]: made round a test-global

File: acceptance/chaos_test.go (hunk 3 above)
Comment (BramGruneir) [inline acceptance/chaos_test.go]: done.  I really wish it would let me do each of the lines of the for on a separate line.

File: acceptance/chaos_test.go [hunk 5]
```diff
@@ -97,6 +104,8 @@ func TestChaos(t *testing.T) {
 	defer func() {
 		<-teardown
 		for i := range clients {
+			clients[i].RLock()
+			defer clients[i].RUnlock()
```
Comment (tschottdorf) [inline acceptance/chaos_test.go]: this will look better two lines down and without the `defer`.

File: acceptance/chaos_test.go [hunk 6]
```diff
@@ -107,33 +116,47 @@ func TestChaos(t *testing.T) {
...
 			for i := 0; i < num; i++ {
 				clients[i].Lock()
 			}
-			for i := 0; i < num; i++ {
-				log.Infof("restarting %v", i)
+			for _, i := range nodes {
+				// Two early exit conditions.
+				select {
+				case <-stopper:
+					break
```
Comment (tschottdorf) [inline acceptance/chaos_test.go]: you're holding the client lock.

Comment (BramGruneir) [inline acceptance/chaos_test.go]: I don't see how.
1) all clients are locked
2) selection of clients are restarted
3) all clients are unlocked

The break will only break the inner loop of 2, so all clients will unlock.
But I can remove the early exit conditions since the kill/restart/init isn't stopping the test from shutting down, the puts are.

Comment (tschottdorf) [inline acceptance/chaos_test.go]: sorry, my bad. Fine as is.

**Comment by tschottdorf** [comment]
LGTM

File: acceptance/chaos_test.go (hunk 5 above)
Comment (BramGruneir) [inline acceptance/chaos_test.go]: done
